### PR TITLE
Add an `Array` sort in the new SMT backend

### PR DESCRIPTION
**Pull request by NlightNFotis**
Along with that, add unit tests the sort and its printing.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Enrico Steffinlongo <enrico.steffinlongo@diffblue.com>

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/solvers/smt2_incremental/smt_sorts.cpp [hunk 1]
```diff
@@ -83,4 +130,9 @@ TEST_CASE("smt_sortt equality", "[core][smt2_incremental]")
   CHECK(bit_vector8 != bool_sort1);
   const smt_bit_vector_sortt bit_vector16{16};
   CHECK(bit_vector8 != bit_vector16);
+  const smt_array_sortt array_sort{smt_bool_sortt{}, smt_bool_sortt{}};
+  CHECK(array_sort != bool_sort1);
+  CHECK(array_sort != bit_vector8);
+  const smt_array_sortt array_sort_bv{bit_vector16, bit_vector16};
+  CHECK(array_sort_bv != array_sort);
```
Comment (TGWDB) [inline unit/solvers/smt2_incremental/smt_sorts.cpp]: Also need some equality tests, including of different argument sorts (and recursive array of array sorts)?

File: unit/solvers/smt2_incremental/smt_to_smt2_string.cpp [hunk 2]
```diff
@@ -21,6 +21,9 @@ TEST_CASE("Test smt_sortt to string conversion", "[core][smt2_incremental]")
 {
   CHECK(smt_to_smt2_string(smt_bool_sortt{}) == "Bool");
   CHECK(smt_to_smt2_string(smt_bit_vector_sortt{16}) == "(_ BitVec 16)");
+  CHECK(
+    smt_to_smt2_string(smt_array_sortt{smt_bool_sortt{}, smt_bool_sortt{}}) ==
+    "(Array Bool Bool)");
```
Comment (TGWDB) [inline unit/solvers/smt2_incremental/smt_to_smt2_string.cpp:26]: Good to check more than one single sort here, recursive and bitvector sorts would be good to check.

File: src/solvers/smt2_incremental/smt_sorts.h [hunk 3]
```diff
@@ -87,6 +87,20 @@ class smt_bit_vector_sortt final : public smt_sortt
   std::size_t bit_width() const;
 };
 
+class smt_array_sortt final : public smt_sortt
+{
+private:
+  const std::string index_sort_label{"index_sort"};
+  const std::string element_sort_label{"element_sort"};
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.h]: :no_entry_sign: You cannot add member fields to an irept, because they will be sliced away for upcast/copy operations. I suggest you reuse the preexisting `ID_index` and `ID_value` `irep_idt`s instead.

File: src/solvers/smt2_incremental/smt_sorts.cpp [hunk 4]
```diff
@@ -62,6 +62,29 @@ std::size_t smt_bit_vector_sortt::bit_width() const
...
+{
+  add(index_sort_label, index_sort);
+  add(element_sort_label, element_sort);
+}
+
+const smt_sortt &smt_array_sortt::get_index_sort() const
+{
+  const irept &index_info = find(index_sort_label);
+  const smt_sortt &index_sort = static_cast<const smt_sortt &>(index_info);
+  return index_sort;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.cpp]: :pick: Spreading this across 3 lines seems a little verbose. You could just write `return static_cast<const smt_sortt &>(find(index_sort_label));`

File: src/solvers/smt2_incremental/smt_sorts.cpp [hunk 5]
```diff
@@ -62,6 +62,29 @@ std::size_t smt_bit_vector_sortt::bit_width() const
...
+}
+
+const smt_sortt &smt_array_sortt::get_element_sort() const
+{
+  const irept &element_info = find(element_sort_label);
+  const smt_sortt &element_sort = static_cast<const smt_sortt &>(element_info);
+  return element_sort;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.cpp]: :pick: This function could also be 1 line instead of 3.

File: unit/solvers/smt2_incremental/smt_sorts.cpp [hunk 6]
```diff
@@ -66,10 +74,54 @@ TEST_CASE("Visiting smt_bit_vec_sortt.", "[core][smt2_incremental]")
...
+
+    void visit(const smt_bool_sortt &) override
+    {
+      bool_visited = true;
+    }
+
+    void visit(const smt_bit_vector_sortt &bit_vec) override
+    {
+      bit_vec_visited = true;
+      REQUIRE(bit_vec.bit_width() == 8);
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/smt_sorts.cpp]: :no_entry_sign: This requirement should be removed as you we do not expect to visit a bit vector sort in this test.

File: unit/solvers/smt2_incremental/smt_sorts.cpp [hunk 7]
```diff
@@ -66,10 +74,54 @@ TEST_CASE("Visiting smt_bit_vec_sortt.", "[core][smt2_incremental]")
...
+      array_visited = true;
+    }
+  } visitor, visitor_second;
+
+  smt_array_sortt{smt_bool_sortt{}, smt_bool_sortt{}}.accept(visitor);
+  REQUIRE_FALSE(visitor.bool_visited);
+  REQUIRE_FALSE(visitor.bit_vec_visited);
+  REQUIRE(visitor.array_visited);
+
+  smt_array_sortt{smt_bit_vector_sortt{64}, smt_bit_vector_sortt{32}}.accept(
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/smt_sorts.cpp:119]: :bulb: I suggest you use `GENERATE` or `GENERATE_REF` as needed to duplicate the test with differing index and element sorts. That way you don't need to duplicate the requirements for each test.

File: src/solvers/smt2_incremental/smt_sorts.cpp [hunk 8]
```diff
@@ -62,6 +62,29 @@ std::size_t smt_bit_vector_sortt::bit_width() const
...
+smt_array_sortt::smt_array_sortt(
+  const smt_sortt &index_sort,
+  const smt_sortt &element_sort)
+  : smt_sortt{ID_smt_array_sort}
+{
+  add(index_sort_label, index_sort);
+  add(element_sort_label, element_sort);
+}
+
+const smt_sortt &smt_array_sortt::get_index_sort() const
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.cpp]: :no_entry_sign: `smt_array_sortt::get_index_sort` should be unit tested like `smt_bit_vector_sortt::bit_width` is.

File: src/solvers/smt2_incremental/smt_sorts.cpp [hunk 9]
```diff
@@ -62,6 +62,29 @@ std::size_t smt_bit_vector_sortt::bit_width() const
...
+{
+  const irept &index_info = find(index_sort_label);
+  const smt_sortt &index_sort = static_cast<const smt_sortt &>(index_info);
+  return index_sort;
+}
+
+const smt_sortt &smt_array_sortt::get_element_sort() const
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.cpp]: :no_entry_sign: `smt_array_sortt::get_element_sort` should be unit tested like `smt_bit_vector_sortt::bit_width` is.

File: src/solvers/smt2_incremental/smt_to_smt2_string.cpp [hunk 10]
```diff
@@ -76,6 +76,12 @@ class smt_sort_output_visitort : public smt_sort_const_downcast_visitort
   {
     os << "(_ BitVec " << bit_vec.bit_width() << ")";
   }
+
+  void visit(const smt_array_sortt &array) override
+  {
+    os << "(Array " << smt_to_smt2_string(array.get_index_sort()) << " "
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_to_smt2_string.cpp]: :hammer: Using `smt_to_smt2_string` will result in constructing and deconstructing additional temporary string streams. The `<<` operator is overridden with an implementation which just uses the existing stream. Therefore `os << "(Array " << array.get_index_sort() << " " << array.get_element_sort() << ")";` is both more succinct and more efficient.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 11]
```diff
@@ -89,6 +89,12 @@ static smt_sortt convert_type_to_smt_sort(const bitvector_typet &type)
   return smt_bit_vector_sortt{type.get_width()};
 }
 
+static smt_sortt convert_type_to_smt_sort(const array_typet &type)
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :no_entry_sign: This additional overload of `convert_type_to_smt_sort` is not required to add the new sort. Therefore it can be separated into a separate commit or even PR.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 12]
```diff
@@ -99,6 +105,10 @@ smt_sortt convert_type_to_smt_sort(const typet &type)
   {
     return convert_type_to_smt_sort(*bitvector_type);
   }
+  if(const auto array_type = type_try_dynamic_cast<array_typet>(type))
+  {
+    return convert_type_to_smt_sort(*array_type);
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: `convert_type_to_smt_sort` is separately unit tested. It would be good to add a new section to the unit test in order to check that the expected `UNIMPLEMENTED_FEATURE` is reached.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

**Comment by NlightNFotis** [comment]
@thomasspriggs @TGWDB We have now addressed all of your concerns, could you please have another look at this PR?

File: src/solvers/smt2_incremental/smt_sorts.cpp [hunk 13]
```diff
@@ -70,7 +89,6 @@ void accept(const smt_sortt &sort, const irep_idt &id, visitort &&visitor)
     return visitor.visit(static_cast<const smt_##the_id##_sortt &>(sort));
 #include "smt_sorts.def"
 #undef SORT_ID
-  UNREACHABLE;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.cpp]: :no_entry_sign: This should not be deleted as it is indeed unreachable for well-formed sorts.

Comment (NlightNFotis) [reply src/solvers/smt2_incremental/smt_sorts.cpp]: Oh yes, sorry this was a mistake - will re-add it ASAP.

File: src/solvers/smt2_incremental/smt_sorts.h (hunk 3 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_sorts.h]: Looks like this is sorted.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 11 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: This has been resolved by removing it from this PR.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 12 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: This has been resolved by removing it from this PR.

File: src/solvers/smt2_incremental/smt_sorts.h [hunk 14]
```diff
@@ -87,6 +87,16 @@ class smt_bit_vector_sortt final : public smt_sortt
...
 };
 
+class smt_array_sortt final : public smt_sortt
+{
+public:
+  explicit smt_array_sortt(
+    const smt_sortt &index_sort,
+    const smt_sortt &element_sort);
+  const smt_sortt &get_index_sort() const;
+  const smt_sortt &get_element_sort() const;
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_sorts.h]: Could you rename `get_index_sort` and `get_element_sort` to `element_sort` and `element_sort` respectively? This would make it more consistent with the naming conventions used for getters elsewhere such as `smt_bit_vector_sortt::bit_width`. This comment should have been in my initial review, but it went awol somewhere, sorry.

**Review by TGWDB — APPROVED** [review]

**Review by thomasspriggs — APPROVED** [review]
